### ISSUE TITLE
[ENH]: move .matplotlib folder from %USERPROFILE% on Windows

### ISSUE THREAD

**Issue by TheOnlyRealCoder** [matplotlib/matplotlib#20779]
### Problem

The folder contains "fontlist-v330.json" in my case.

### Proposed solution

Move the directory to a subdirectory in %APPDATA%, where it belongs (as the name suggests).

**Comment by tacaswell** [comment]
Do you have a reference for what "should" be in various places?

I suspect that this would be addressed by https://github.com/matplotlib/matplotlib/issues/12549 , but if we do not want to pick up that dependency, the place in the code to make the change is https://github.com/matplotlib/matplotlib/blob/e291cdbbf1bfb58415f30be6d2217350248e7dec/lib/matplotlib/__init__.py#L470-L478 

Even though this is an internal cache, we should document that we are moving it on windows (as I'm sure there exist other users who have discovered where it lives and are either deleting it (for reasons) or pre-populating it (again, for reasons).

**Comment by tacaswell** [comment]
Milestoned for 3.6, but if it is ready it can go in for 3.5.  Labeled good first issue because it only requires changing details of how we discover the cache directory and should not touch our public API (other than a sideeffect of where files go), but may require intermediate Python (as I think we want to generalize `_get_xgd_cache_dir` to also have a windows path) and familiarity with Windows.

**Comment by TheOnlyRealCoder** [comment]
> 
> 
> Do you have a reference for what "should" be in various places?
> 
https://docs.microsoft.com/en-us/windows/apps/design/app-settings/store-and-retrieve-app-data

**Comment by NeilGirdhar** [comment]
> (as I think we want to generalize _get_xgd_cache_dir to also have a windows path) and familiarity with Windows.

Why not just use `appdirs` as suggested in #12549?

**Comment by StefRe** [comment]
> Move the directory to a subdirectory in %APPDATA%, where it belongs (as the name suggests).

I think it should go into **%LOCALAPPDATA%**, not %APPDATA% which is for roaming app data. Cached fonts data are certainly not applicable on other devices. See also https://docs.microsoft.com/en-us/windows/win32/shell/knownfolderid.

**Comment by StefRe** [comment]
On Windows, the `get_configdir` and `get_cachedir` currently return the same value:
https://github.com/matplotlib/matplotlib/blob/c072e3a800347de225577ef6692e500aaec294de/lib/matplotlib/__init__.py#L467-L472

On Windows, I think we need the to differentiate here between %APPDATA% for roaming config data (get_configdir) and %LOCALAPPDATA% for local cache data (get_cachedir).

**Comment by NeilGirdhar** [comment]
@StefRe I think it should go into %LOCALAPPDATA%, not %APPDATA% which is for roaming app data. 

appdirs already figures this out for you if you pass `roaming=True`: https://github.com/ActiveState/appdirs/blob/8eacfa312d77aba28d483fbfb6f6fc54099622be/appdirs.py#L165

**Comment by StefRe** [comment]
@NeilGirdhar  Thanks for mentioning, but all I wanted to say was just what files should go to what folders, not how to obtain these folder locations.

**Comment by StefRe** [comment]
@tacaswell If we want to fix this particular issue #20779 for Windows, I can make a PR.   
(Switching to `appdirs `#12549 is beyond me as I have no masOS or FreeBSD to test it and update the documention of [`get_configdir`](https://matplotlib.org/stable/api/matplotlib_configuration_api.html#matplotlib.get_configdir) and [`get_cachedir`](https://matplotlib.org/stable/api/matplotlib_configuration_api.html#matplotlib.get_cachedir). I couldn't find docs for `appdirs` (apart from the [README](https://github.com/ActiveState/appdirs/blob/master/README.rst) file) that we could refer to for the details about folder selection on different OSs.)

**Comment by NeilGirdhar** [comment]
@StefRe Yes, but I understand, but if you're going to fix this, I really think using appdirs is the right approach. I can test it for you on macOS.  Otherwise, you'll just be fixing it for one OS for now, and it could easily break again one day.

**Comment by StefRe** [comment]
@NeilGirdhar I didn't say that it's hard to use `appdirs` (quite the contrary), I just see the documentation issue as I don't trust myself to properly document folder locations for the different operating systems in the matplotlib documentation when using `appdirs`. Right now it's easy: we only have three cases to document: Linux, Windows and all other. I don't know on what different Unix or macOS systems matplotlib can be run and what `appdirs` will return in each single case (ok, I could check the source and document it myself, but it would be much easier if they had some authorative documentation one could refer to). So it's not that I wouldn't like `appdirs`, I'm just feeling a bit uncomfortable relying on a third party programm without proper documentation (at least I couldn't find it).
<br>  

>  Just call appdirs.user_data_dir(appname='matplotlib', appauthor='matplotlib', version=..., roaming=False) ...

I'd use `...\matplotlib` instead of `...\matplotlib\matplotlib` as do for instance Python, Spyder, Gimp and other programs that don't have a company as an author (i.e. set `appauthor=''` in the call to `user_data_dir`)

**Comment by NeilGirdhar** [comment]
> I'd use ...\matplotlib instead of ...\matplotlib\matplotlib as do for instance Python, Spyder, Gimp and other programs that don't have a company as an author (i.e. set appauthor='' in the call to user_data_dir)

You're right!  Good point.

> I'm just feeling a bit uncomfortable relying on a third party programm without proper documentation (at least I couldn't find it).

I understand.  That's fair.  But the issue with just hacking a custom solution is that it remains messed up for the rest of us :/  I didn't think about the documentation issue.

**Comment by jkseppan** [comment]
appdirs is now deprecated; the replacement is [platformdirs](https://github.com/tox-dev/platformdirs)

**Comment by VedantMadane** [comment]
Hi, I'd like to work on this issue. I'll implement the change to use \platformdirs\ (the maintained replacement for \ppdirs\) to properly place the matplotlib configuration and cache directories on Windows (%LOCALAPPDATA% for cache, %APPDATA% for config if roaming is desired, otherwise %LOCALAPPDATA%).

**Comment by NeilGirdhar** [comment]
@timhoffm Could you please reopen the issue?  This is only fixed on Windows, and in a very brittle way. Or perhaps we should open a separate issue for relying on platformdirs?
